sdk/src/types/block/semantic/state_transition.rs

```diff
@@ -213,6 +226,74 @@ impl StateTransitionVerifier for AccountOutput {
             _ => {}
         }
 
+        match (current_state.features().staking(), next_state.features().staking()) {
+            (None, Some(staking_output)) => {
+                let past_bounded_slot_index = context
+                    .protocol_parameters
+                    .past_bounded_slot(context.commitment_context_input.unwrap());
+                let past_bounded_epoch_index = context.protocol_parameters.epoch_index_of(past_bounded_slot_index);
+
+                if staking_output.start_epoch() != past_bounded_epoch_index {
+                    return Err(TransactionFailureReason::StakingStartEpochInvalid);
+                }
+                if staking_output.end_epoch()
+                    < past_bounded_epoch_index + context.protocol_parameters.staking_unbonding_period
+                {
+                    return Err(TransactionFailureReason::StakingEndEpochTooEarly);
+                }
+            }
+            (Some(staking_input), None) => {
+                let future_bounded_slot_index = context
+                    .protocol_parameters
+                    .future_bounded_slot(context.commitment_context_input.unwrap());
+                let future_bounded_epoch_index = context.protocol_parameters.epoch_index_of(future_bounded_slot_index);
+
+                if staking_input.end_epoch() >= future_bounded_epoch_index {
+                    return Err(TransactionFailureReason::StakingFeatureRemovedBeforeUnbonding);
+                } else if !context.mana_rewards.contains_key(current_output_id)
+                    || !context.reward_context_inputs.contains_key(current_output_id)
+                {
+                    return Err(TransactionFailureReason::StakingRewardClaimingInvalid);
+                }
+            }
+            (Some(staking_input), Some(staking_output)) => {
+                let past_bounded_slot_index = context
+                    .protocol_parameters
+                    .past_bounded_slot(context.commitment_context_input.unwrap());
+                let past_bounded_epoch_index = context.protocol_parameters.epoch_index_of(past_bounded_slot_index);
+                let future_bounded_slot_index = context
+                    .protocol_parameters
+                    .future_bounded_slot(context.commitment_context_input.unwrap());
+                let future_bounded_epoch_index = context.protocol_parameters.epoch_index_of(future_bounded_slot_index);
+
+                if staking_input.end_epoch() >= future_bounded_epoch_index {
```

Comment:
Why not the other way around as in the TIP?

Reply:
Felt like it's more consistent with the rest of our checks that way, I can reverse if you prefer

Reply:
I thought our checks are consistent with the TIP :trollface: Don't care that much, feel free to leave as it is